### PR TITLE
fix(health): verify APCu with a store/fetch round trip, and fix the CI ini typo (#835)

### LINKED ISSUE #835: APCu is misconfigured in CI and its availability check would select a cache that silently stores nothing
Two related APCu problems, filed together because the same fact settles both: **production uses Redis, and APCu is deliberately off under CLI.**

## Evidence

Measured on the deployment host, in the PHP CLI binary the WebSocket server runs under:

| | |
|---------------------------------------------|--------------------------------|
| `extension_loaded('apcu')`                  | `true`                         |
| `apc.enabled`                               | `1`                            |
| `apc.enable_cli`                            | **`0`**                        |
| `extension_loaded('redis')`                 | `true`                         |
| `Health` backend selection                  | tries **Redis first**; APCu only if Redis is absent |

So the live WebSocket server caches through Redis and never touches APCu.

---

## 1. The CI ini setting is a no-op (typo)

`.github/workflows/phpunit.yml` sets:

```yaml
ini-values: apcu.enable_cli=1, opcache.enable=1, opcache.enable_cli=1
```

**APCu's setting is `apc.enable_cli`, not `apcu.`** — it inherited the prefix from APC. So the line does nothing: the extension loads (making `function_exists('apcu_store')` true) while CLI support stays off, so `apcu_store()` silently no-ops and `apcu_exists()` returns false.

That combination is unusually good at hiding: any test guarding on `function_exists()` concludes APCu is present, then finds it does not work. It surfaced in #834, where a shim guarded with `!function_exists('apcu_store')` correctly stood down in CI and left the tests talking to a dead extension.

### Fix: correct the name, and set it to `0`

Not `1`. Enabling it would:

- exercise a caching path **production never takes**, since production selects Redis;
- introduce **process-wide state across 3116 tests** — APCu persists for the life of the PHP process, so one test caching a response changes what a later test sees, which is a classic source of order-dependent flakiness;
- buy no coverage that corresponds to any real configuration.

The cache *logic* is better covered deterministically by a shim (the approach #834 takes) than by a real backend whose state leaks between tests.

The current behaviour is therefore already correct — but correct **by accident**. Setting `apc.enable_cli=0` under the right name turns a typo into a recorded decision, with a comment noting that production uses Redis and APCu is deliberately off under CLI.

---

## 2. The availability check would select a cache that stores nothing

`src/Health.php` decides the backend with:

```php
$apcuAvailable = extension_loaded('apcu')
    && function_exists('apcu_exists')
    && function_exists('apcu_store')
    && function_exists('apcu_fetch');
```

It never calls `apcu_enabled()`. **On production right now, all four are true while `apc.enable_cli=0`.** Redis is up, so this is not currently reached — but if Redis became unavailable, the fallback would select APCu and get a cache that silently stores nothing: every `cacheSet()` reports success, every `cacheGet()` misses, and the only symptom is quiet slowness.

This is the same defect family as #822 and #833 — a check reporting success it never established — and it is latent in exactly the worst place: the path taken *during* a Redis outage, when the fallback is what you are relying on.

### Fix

Add `apcu_enabled()` to the predicate (it exists whenever the extension does, and returns false when disabled or CLI-disabled). Both sites need it — the initial selection and the Redis-connection-lost fallback, which duplicate the check.

Better still, verify with a round-trip: `apcu_store()` a probe key and read it back. `apcu_enabled()` answers the configuration question; a round-trip answers the one that matters, and would also catch a full or broken segment. Whichever is chosen, when the check fails the backend must fall to `none` with `cacheEnabled = false`, exactly as it does when the extension is absent.

A test asserting that an enabled-but-non-functional APCu is **not** selected would keep this closed. Neither existing check is exercised by any test today, which is why the gap survived.

---

Found while diagnosing CI failures on #834. Neither problem is caused by that PR, and neither should be fixed inside it — the workflow change in particular has suite-wide blast radius and deserves to land on its own.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### PR DESCRIPTION
**Pull request by JohnRDOrazio**
Closes #835.

Two related APCu problems, fixed as two commits because the workflow change has suite-wide blast radius and deserves to be revertable on its own.

## 1. The CI ini setting was a no-op (typo)

`.github/workflows/phpunit.yml` set `apcu.enable_cli=1`. APCu inherited APC's prefix, so the real name is `apc.enable_cli` — the line did nothing. It is now `apc.enable_cli=0`, **not** `1`, with a comment recording why: production selects Redis and never takes the APCu path, and APCu state persists for the life of the PHP process, so a real APCu would leak one test's cache entry into what a later test sees across the whole suite. The cache logic is covered deterministically by `phpunit_tests/Support/ApcuShim.php` instead.

The resulting behaviour is unchanged — it was already correct *by accident*. This turns it into a recorded decision.

## 2. The availability check would have selected a cache that stores nothing

The old predicate — `extension_loaded('apcu')` plus `function_exists()` on three `apcu_*` functions — is true on the deployment host while `apc.enable_cli=0`. Selecting that backend yields a cache that silently swallows everything: every `cacheSet()` reports success, every `cacheGet()` misses, and the only symptom is quiet slowness. It is latent rather than live only because Redis is up — the path that reaches it is the **Redis-outage fallback**, i.e. precisely when the fallback is what you are relying on.

Both detection sites (initial selection in `onOpen()`, and `handleRedisFailure()`) now call a new `Health::apcuUsable()`, which stores a random probe key and reads it back. A round trip was chosen over `apcu_enabled()` because it also catches a full or broken shared-memory segment, which no configuration flag reports. On failure the backend falls to `none` with `cacheEnabled = false`, exactly as when the extension is absent.

Three deviations from the issue text, each reasoned in the docblock:

- **The existence check resolves namespace-first** (`LiturgicalCalendar\Api\apcu_store` before global), mirroring how PHP resolves the unqualified `apcu_*` calls in that file. This is also what makes the decision testable, via the same mechanism #834's shim uses.
- **`extension_loaded('apcu')` is dropped** as subsumed: an unloaded extension defines no functions, so the existence loop already stands down.
- **The function list grew from three to five.** `cacheDelete()` calls `apcu_delete()` and `cacheInfo()` calls `apcu_sma_info()`; neither was checked, and an undefined function is a fatal, not a miss.

## Tests

New `phpunit_tests/HealthApcuDetectionTest.php` (5 tests): probe-level negative and positive, probe cleanup, and both directions through the real `handleRedisFailure()` site. `ApcuShim` gained `simulateDisabledStore()` (store reports success, stores nothing) and `keys()` — the round trip needs no `apcu_enabled` stub, but *simulating* an inert APCu does need a fault switch.

Pre-fix the new suite reported `Errors: 3, Failures: 1`; post-fix `OK (5 tests, 13 assertions)`.

Honest limitation, also stated in the test docblock: on a host without ext-apcu the *negative* test would pass pre-fix for the wrong reason. It is the positive/negative pair that makes it a real differential.

## Verification

- `composer lint` (phpcs): clean
- `composer analyse` (PHPStan level 10): `[OK] No errors`, 322 files
- `composer test:quick`: exit 0 — `Tests: 2923, Assertions: 177760, Skipped: 11`

## Follow-up

The identical predicate survives in six other places (`src/Utilities.php` ×5, `src/Models/MissalsPath/MissalMetadataMap.php` ×1). Those are on the FPM request path rather than the WebSocket CLI process, so they are not a live defect — filed separately as #836.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved cache reliability by verifying that APCu can store and retrieve data before enabling it.
  * Updated Redis-failure fallback behavior to disable caching when APCu is unavailable, nonfunctional, or encounters errors.
  * Corrected CLI test configuration to disable APCu as intended.

* **Tests**
  * Added coverage for APCu detection, fallback selection, probe cleanup, error handling, and nonpersisting cache behavior.

### CODE REVIEW

**Comment by JohnRDOrazio** [comment]
Thanks — one finding applied, one rejected. Details below.

## Applied: `\Throwable`-safety of the probe (`2b8bac51`)

Valid, and it was two holes rather than one:

- **`random_bytes()` sat outside the `try`** and can throw. `apcuUsable()` owes its callers a bool — a `false` is what makes the backend fall to `none`. A throw would instead propagate out of `onOpen()` or, worse, out of `handleRedisFailure()`, which runs during the very Redis outage the APCu fallback exists to cover. That turns a degraded cache into a broken connection handler.
- **The `finally` was its own hazard.** An exception raised in a `finally` *replaces* whatever the `try` was returning, so an unlucky `apcu_delete()` could mask an answer that had already been decided correctly. Cleanup is now best-effort and cannot change the result.

The `catch` reports before returning `false` rather than swallowing, since an exception reaching it is by definition one nobody anticipated.

Covered by two new tests — the probe itself and the Redis-fallback site — driven by a new `ApcuShimStore::simulateThrowingStore()`. That is deliberately a *separate* fault switch from the existing `simulateDisabledStore()`: one models a backend that answers dishonestly, the other one that blows up.

`composer lint` clean, `composer analyse` (PHPStan level 10) `[OK] No errors`, `composer test:quick` green at `Tests: 2925`.

## Rejected: `#[Group('slow')]` on `HealthApcuDetectionTest`

This one would make things worse, so I have not applied it.

**The test runs in 7 ms** (`Time: 00:00.006, Memory: 18.00 MB` for the full class). Per this repo's `CLAUDE.md`, the `slow` group is *"reserved for tests with **measurable** runtime cost — e.g., multi-year calendar calculations, rate-limit window waits, or full-schema-corpus validation"*, and it explicitly notes that *"integration tests are NOT automatically slow: most `Routes/*` tests run in < 200 ms and are excluded from the `slow` group."*

Adding the attribute would exclude this class from `composer test:quick` — the command developers actually run — removing the regression guard for a latent production defect in order to save 7 ms. The group is an exclusion mechanism, not a label.

### Why this suggestion keeps recurring, for what it's worth

The repo's own `CLAUDE.md` is the likely cause. Line 452 reads, bolded and set off as its own blockquote:

> **Always use the `#[Group('slow')]` ATTRIBUTE, never a legacy `@group slow` docblock.**

Read on its own — which is how a chunked retrieval will read it — that is an unconditional instruction to apply `#[Group('slow')]`. Its actual subject is the *spelling* (attribute vs. docblock, because PHPUnit 12 honours only the attribute), and the *scoping* condition lives in a separate list item further up. The bold-blockquote formatting makes it more likely to be retrieved standalone, not less.

That wording is being fixed, along with a `.coderabbit.yaml` path instruction for `phpunit_tests/**` stating the rule directly, so this stops recurring on every PR that adds a fast test.

File: phpunit_tests/HealthApcuDetectionTest.php [hunk 1]
```diff
@@ -0,0 +1,187 @@
+<?php
+
+declare(strict_types=1);
+
+namespace LiturgicalCalendar\Tests;
+
+use LiturgicalCalendar\Api\ApcuShimStore;
+use LiturgicalCalendar\Api\Health;
+use PHPUnit\Framework\Attributes\CoversClass;
+use PHPUnit\Framework\TestCase;
```
[the remark this answers is not in the document]
Comment (JohnRDOrazio) [reply phpunit_tests/HealthApcuDetectionTest.php:10]: the test runs in 7 ms, and CLAUDE.md reserves slow for measurable runtime cost; marking it slow would drop it out of composer test:quick, deleting the regression guard from the run developers actually use. What's tripping CodeRabbit

Your own CLAUDE.md, line 452 — bolded, and set off as its own standalone blockquote:

▎ Always use the #[Group('slow')] ATTRIBUTE, never a legacy @group slow docblock.

The subject of that sentence is the spelling of the group. But its surface grammar is an unconditional imperative: always use #[Group('slow')]. The condition that actually governs whether the group applies — "reserved for tests with measurable runtime cost" — lives in a separate list item further up, in a different block.

Reworded CLAUDE.md so the default comes first ("most tests do not belong in the group, and adding them to it is a mistake"), the conditional sits inside the imperative sentence, and the sentence says outright that it's about spelling rather than application.
